File: Python3/652. Find Duplicate Subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.pattern = collections.defaultdict(list)
        
        def dfs(node: TreeNode) -> str:
            if not node:
                return 'None'
            
            # leaf node
            if not node.left and not node.right:
                self.pattern[node.val].append(node)
                return str(node.val)
            
            left = dfs(node.left)
            right = dfs(node.right)
            
            ptn = str(node.val) + ',' + left + ',' + right
            
            self.pattern[ptn].append(node)
            #print(ptn)
            
            return ptn
        
        dfs(root)
        ans = []
        
        for nodes in self.pattern.values():
            if len(nodes) > 1:
                ans.append(nodes[0])
                
        return ans

File: Java/652. Find Duplicate Subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<String, List<TreeNode>> patterns;
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        patterns = new HashMap<>();
        dfs(root);
        
        List<TreeNode> ans = new ArrayList<>();
        for(List<TreeNode> nodes: patterns.values()){
            if(nodes.size() > 1){
                ans.add(nodes.get(0));
            }
        }
        
        return ans;
    }
    
    private String dfs(TreeNode node){
        if(node == null){
            return "null";
        }
        
        String ptn = Integer.toString(node.val);
        
        String left = dfs(node.left);
        String right = dfs(node.right);
        
        if(node.left == null && node.right == null){
            ;
        }else{
            ptn += "," + left + "," + right;
        }
        
        if(!patterns.containsKey(ptn)){
            patterns.put(ptn, new ArrayList<TreeNode>());
        }
        patterns.get(ptn).add(node);
        return ptn;
    }
}